src/tokens/reader_stream.rs

```diff
@@ -0,0 +1,560 @@
+const INVALID_TOKEN_PANIC_TEXT: &str = "Invalid underlying reader state";
+
+// TODO make this more generic with respect to other readers--the problem is Item/Symbol
+// TODO this has to abstract over potentially system reader to implement macros
+
+/// Internal tracking state of the returned token.  This is necessary because the lifetime of the
+/// token returned from the stream is bound to the stream itself.
+///
+/// This is essentially strongly typed [`Option`] with convenience to operate within the
+/// context of a [`ReaderTokenStream`]
+enum UnderlyingReader<R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol>,
+{
+    Active(R),
+    Inactive,
+}
+
+impl<R> UnderlyingReader<R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol>,
+{
+    fn try_mut_ref(&mut self) -> IonResult<&mut R> {
+        use UnderlyingReader::*;
+        match self {
+            Active(reader) => Ok(reader),
+            Inactive => illegal_operation(INVALID_TOKEN_ERR_TEXT),
+        }
+    }
+
+    fn try_ref(&self) -> IonResult<&R> {
+        use UnderlyingReader::*;
+        match self {
+            Active(reader) => Ok(reader),
+            Inactive => illegal_operation(INVALID_TOKEN_ERR_TEXT),
+        }
+    }
+
+    fn as_mut_ref(&mut self) -> &mut R {
+        self.try_mut_ref().expect(INVALID_TOKEN_PANIC_TEXT)
+    }
+
+    fn as_ref(&self) -> &R {
+        self.try_ref().expect(INVALID_TOKEN_PANIC_TEXT)
+    }
+
+    /// Invalidates this reader returning its original underlying reader.
+    fn invalidate(&mut self) -> Self {
+        std::mem::replace(self, UnderlyingReader::Inactive)
+    }
+}
+
+/// Indicates the state of the last token returned.
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+enum PendingToken {
+    Materialized,
+    Deferred,
+    None,
+}
+
+impl PendingToken {
+    fn is_deferred(&self) -> bool {
+        matches!(self, PendingToken::Deferred)
+    }
+}
+
+/// Condensed version of the item of the [`IonReader`].
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+enum ReaderPosition {
+    /// We are positioned on nothing.
+    Nothing,
+    /// We are positioned on some value.
+    NonContainer,
+    /// We are positioned on the start of some container.
+    Container(ContainerType),
+}
+
+/// Adapter for a [`TokenStream`] over an arbitrary [`IonReader`]
+pub struct ReaderTokenStream<'a, R>
+where
+    R: IonReader<Item = StreamItem, Symbol = Symbol> + 'a,
+{
+    // XXX this is so we can have multiple closures to lazily evaluate tokens
+    reader_cell: Rc<RefCell<UnderlyingReader<R>>>,
+    pending_token: PendingToken,
+    position: ReaderPosition,
+    // XXX this allows us to explicitly capture a lifetime for the reader we operate on
+    phantom: PhantomData<&'a R>,
+}
+
+/// These generate the methods that produce lazy content for our stream against the underlying
+/// reader.  There is a lot of boilerplate and variants so this avoids copy/paste errors.
+macro_rules! scalar_content_method {
```

Comment:
🗺️ It makes me sad to do this, but it does eliminate *a lot* of boilerplate for generating the `Content` instances against the `Rc<RefCell<...>>`.